src/DynamoCore/Graph/Nodes/NodeModel.cs

```diff
@@ -1717,15 +1717,15 @@ public void Warning(string p, bool isPersistent = false)
                 State = ElementState.PersistentWarning;
                 if (!string.Equals(persistentWarning, p))
                 {
-                    persistentWarning += p;
+                    persistentWarning += persistentWarning == "" ? p : $"{Environment.NewLine}{p}";
                 }
                 ToolTipText = persistentWarning;
             }
             else
             {
                 State = ElementState.Warning;
                 transientWarning = p;
-                ToolTipText = string.IsNullOrEmpty(persistentWarning) ? p : string.IsNullOrEmpty(p) ? persistentWarning : $"{persistentWarning}\n{p}";
+                ToolTipText = string.IsNullOrEmpty(persistentWarning) ? p : string.IsNullOrEmpty(p) ? persistentWarning : $"{persistentWarning}{Environment.NewLine}{p}";
```

Comment:
Probably a larger task...but we should redesign the data here to somthing like a HashSet or a List
I would say there should only be a single Warnings enumerable. Each warning should have a UniqueId, Content, type/category, persistent/transient.
The Tooltip should be a UI property, I think, and it should be generated from the existing Warnings enumerable.

Reply:
Agreed, some mature design should be in favor. Especially our team discussed per port and node warning system.

Reply:
@QilongTang the per port warning idea seems quite hard. I was discussing it with @filipeotero and was looking at say passing 3 nulls to the x, y, z inputs for a Point.ByCoordinates node. The node fails at the first null and bails so it seems unlikely that we can get it to warn for any subsequent invalid inputs.